posts/2024-02-15-waku-network-tech-overview.mdx

```diff
@@ -0,0 +1,178 @@
+
+Autosharding is a simple protocol that dispatches messages on a shard based on the application.
+
+### Message Routing - RLN Relay
+
+Waku is generalized, meaning any kind of payload can be transported. Hence, there is no strict definition of “spam”. Whether the message contains a [meme](https://github.com/b4s36t4/waku-meme) or a [zk note for private DeFi](https://railgun.ch/#/), Waku shouldn't even know the content; metadata and data should remain private.
+
+Hence, there is a risk of someone pushing gigabits of data to the network. This can be a problem at different levels:
+
+- Bandwidth usage: this may lead to a user’s bandwidth being hogged and affecting other services (streaming, staking), or having a surprise bill.
+- Connectivity: if a node does not have enough bandwidth to send/receive all messages on the gossipsub layer, then their behavior may be seen as incorrect from other nodes that may disconnect from it.
+- Reliability: If the traffic is higher than the available bandwidth, then a node may not be able to reliably send and receive messages.
+- Other resources: memory usage is correlated with the traffic, so is disk space for a store service.
+
+So rather than defining what a "spam" message looks like, we introduced rate limit on the network to enable a fair use of Waku, with a capping of the bandwidth usage per shard.
+
+This is done with RLN, or Rate Limiting Nullifier, which enables to limit the rate of messages sent by a given publisher.
+It is currently set to 1 msg/s. Coupled with a maximum message size (150kB) and a maximum number of publishers (80k, TBD), we are able to assume a maximum bandwidth usage per shard (around 10Mbps).
+
+Rate-limiting publishers in a censorship-resistant and private manner is difficult; this is why we are using zero-knowledge technology:
+
+1. Users push their RLN credentials to a smart contract (currently on Ethereum Sepolia testnet).
+2. Nodes watch the contract and include the new membership.
+3. When sending a message, the user attaches an RLN proof to the message with the current epoch (timestamp in seconds).
+4. Nodes can verify the proof, without being able to correlate the user’s Ethereum address (used on the smart contract) with the message being sent, preserving anonymity.
+5. If a user attempts to send more than 1 msg/s, nodes can detect this and drop the message surplus, or spam, protecting the network.
+
+### Servicing Mostly-offline and Resource-restricted Devices
+
+Finally, how is Waku useful to mobile and browsers?
+
+Waku defines a number of [request-reply protocols](https://rfc.vac.dev/spec/10/#requestreply-domain) to enable such devices to access the Waku network, without having to be always online or consume extensive amounts of bandwidth, i.e., without participating in the Waku Relay network.
```

Comment:
*request-response (trying to bring some consistency to our terminology across platforms)

Reply:
I also prefer "request-response" but our specs says "request-reply": https://rfc.vac.dev/spec/10/#requestreply-domain
Very happy to deprecate "request-reply" I think it sounds weird.

Reply:
Yeah, I don't think we've been very consistent, but I'd rather we update RFC 10 to "request-response" (most other RFCs already use "request-response", e.g. Store here: https://rfc.vac.dev/spec/13/#wire-specification)